File: appmap/manifest.py
"""Wholesale manifest.yaml build — the derived 'front door' index (tier-1).

All fields here are derived from the surface records; the manifest is safe to
regenerate at any time. Kept low-stakes on merge by being fully derived: on a
conflict, take either side and re-render.
"""

from __future__ import annotations

import datetime as _dt
import subprocess
from pathlib import Path
from typing import Any

import yaml

from .config import Config
from .links import LinkGraph
from .model import Surface


def _git_sha(root: Path) -> str | None:
    try:
        out = subprocess.run(
            ["git", "-C", str(root), "rev-parse", "--short", "HEAD"],
            capture_output=True, text=True, timeout=5,
        )
        if out.returncode == 0:
            return out.stdout.strip()
    except Exception:
        pass
    return None


def build_manifest(
    surfaces: list[Surface], links: LinkGraph, cfg: Config
) -> dict[str, Any]:
    index: dict[str, Any] = {}
    for s in sorted(surfaces, key=lambda s: s.id):
        lv = s.last_verified.get("sha") if s.last_verified else None
        index[s.id] = {
            "title": s.title,
            "kind": s.kind,
            "path": f"surfaces/{s.id}/surface.md",
            "last_verified": lv,
            "needs_review": s.needs_review,
        }

    review_queue = sorted(s.id for s in surfaces if s.needs_review)

    dangling = [
        {"from": d.from_id, "to": d.to, "kind": d.kind, "ref": d.ref}
        for d in links.dangling
    ]

    manifest: dict[str, Any] = {
        "generated_at": _dt.datetime.now(_dt.timezone.utc)
        .replace(microsecond=0)
        .isoformat(),
        "map_sha": _git_sha(cfg.project_root),
        "launch_surface": cfg.launch_surface,
        "surfaces": index,
        "review_queue": review_queue,
        "link_health": {
            "dangling_links": dangling,
        },
    }
    return manifest


def write_manifest(manifest: dict[str, Any], cfg: Config) -> Path:
    text = yaml.safe_dump(manifest, sort_keys=False, allow_unicode=True)
    cfg.manifest_path.write_text(text, encoding="utf-8")
    return cfg.manifest_path

File: cli/Sources/AppMapKit/Manifest.swift
// Wholesale manifest.yaml build — the derived "front door" index (tier-1).
//
// All fields are derived from the surface records; the manifest is safe to
// regenerate at any time (on a merge conflict, take either side and re-render).
// It deliberately carries no volatile fields (no wall-clock timestamp, no git
// sha of its own) and is emitted by a deterministic hand-rolled YAML writer, so
// a re-render on unchanged records produces a byte-identical file.

import Foundation

public func buildManifestYAML(
    surfaces: [SurfaceRecord], links: LinkGraph, cfg: Config
) -> String {
    var out: [String] = []
    out.append("launch_surface: \(yamlScalar(cfg.launchSurface))")

    if surfaces.isEmpty {
        out.append("surfaces: {}")
    } else {
        out.append("surfaces:")
        for s in surfaces.sorted(by: { $0.id < $1.id }) {
            out.append("  \(yamlScalar(s.id)):")
            out.append("    title: \(yamlScalar(s.title))")
            out.append("    kind: \(yamlScalar(s.kind))")
            out.append("    path: \(yamlScalar("surfaces/\(s.id)/surface.md"))")
            out.append("    last_verified: \(yamlScalar(s.lastVerified["sha"] as? String))")
            out.append("    needs_review: \(s.needsReview)")
        }
    }

    let reviewQueue = surfaces.filter(\.needsReview).map(\.id).sorted()
    if reviewQueue.isEmpty {
        out.append("review_queue: []")
    } else {
        out.append("review_queue:")
        for id in reviewQueue { out.append("- \(yamlScalar(id))") }
    }

    out.append("link_health:")
    if links.dangling.isEmpty {
        out.append("  dangling_links: []")
    } else {
        out.append("  dangling_links:")
        for d in links.dangling {
            out.append("  - from: \(yamlScalar(d.fromID))")
            out.append("    to: \(yamlScalar(d.to))")
            out.append("    kind: \(yamlScalar(d.kind))")
            out.append("    ref: \(yamlScalar(d.ref))")
        }
    }

    return out.joined(separator: "\n") + "\n"
}

/// Emit one YAML scalar: plain when unambiguous, single-quoted otherwise,
/// `null` for nil. Deterministic by construction.
func yamlScalar(_ value: String?) -> String {
    guard let value else { return "null" }
    if value.isEmpty { return "''" }
    let plainSafe = value.range(
        of: "^[A-Za-z0-9][A-Za-z0-9 ._/()-]*$", options: .regularExpression
    ) != nil
    let ambiguous = [
        "null", "~", "true", "false", "yes", "no", "on", "off",
    ].contains(value.lowercased())
        || value.range(of: "^[0-9.+-]", options: .regularExpression) != nil
        || value.hasSuffix(" ")
    if plainSafe && !ambiguous { return value }
    return "'" + value.replacingOccurrences(of: "'", with: "''") + "'"
}
